Add documentation for SSL truststore issue in Hive (#220)
exasol/public-knowledgebase: Connect-with-Exasol/hive-virtual-schema-with-ssl-fails-with-no-such-file-or-directory-for-ssltruststore.md

## Symptoms

A Hive JDBC connection works successfully without SSL. After SSL is enabled, the following behavior can be observed:

- `CREATE CONNECTION` succeeds.
- `CREATE VIRTUAL SCHEMA` succeeds.
- Tables are visible in the Virtual Schema.
- Queries against the Virtual Schema fail.

Example:

```sql
CREATE OR REPLACE CONNECTION HIVE_SSL_CONN
TO 'jdbc:hive2://hive.example.com:10001/;
SSL=1;
transportMode=http;
httpPath=cliservice;
SSLTrustStore=/buckets/bfsdefault/buckethive/example-truststore.jks;
SSLTrustStorePwd=changeit'
USER 'example_user'
IDENTIFIED BY 'example_password';

CREATE VIRTUAL SCHEMA HIVE_VS
USING ADAPTER.HIVE_JDBC_ADAPTER
WITH
    CONNECTION_NAME = 'HIVE_SSL_CONN'
    SCHEMA_NAME = 'example_schema';
```

The Virtual Schema is created successfully, but querying a table returns an error similar to:

```text
ETL-5402: JDBC-Client-Error:
Connecting to 'jdbc:hive2://...'
failed:

[Cloudera][HiveJDBCDriver](500164)
Error initialized or created transport for authentication:

/buckets/bfsdefault/buckethive/example-truststore.jks

(No such file or directory)
```

## Cause

The error is **not** caused by SSL itself.

> Instead, the Hive JDBC driver cannot access the truststore file referenced in the JDBC URL from the runtime environment where the query is executed.

Reviewer: ```suggestion
Instead, the Hive JDBC driver used by the ExaLoader cannot access the truststore file referenced in the JDBC URL from the runtime environment where the query is executed.
```

> When referencing files in Exasol, there are two different views of BucketFS.

Reviewer: ```suggestion
When referencing files in Exasol, there are two different views of the filesystem behind BucketFS. One of them is a change-root used by the UDFs.

That means that the Virtual Schema Adapter (a UDF) and the ExaLoader (running in COS) never see the same paths.
```

Reviewer: The JDBC driver comes into play here twice. Once inside the VS adapter, where the path in the JDBC URL works and a second time in the ExaLoader, that runs in the global COS context. And that second use is where the BucketFS path does not fit.

### Logical BucketFS path

This is the path used by UDFs and adapter scripts.

```text
/buckets/bfsdefault/buckethive/example-truststore.jks

> ```

Reviewer: This is a change-root ([chroot](https://en.wikipedia.org/wiki/Chroot)) environment, meant to prevent UDFs from accessing any files of the underlying operating system. This is an intentional security mechanism.

Author: rewritten, I added content chroot

### Physical BucketFS location

> This is the location where the file is actually stored on disk.

Reviewer: ```suggestion
This is absolute path of the truststore file in the COS filesystem context..
```

```text
/exa/data/bfsdefault/buckethive/example-truststore.jks
```

The Hive JDBC connection references the logical BucketFS path:

```text
SSLTrustStore=/buckets/bfsdefault/buckethive/example-truststore.jks
```

The reported error

```text
No such file or directory
```

> indicates that the JDBC driver successfully parsed the connection string but could not open the truststore because the specified path does not exist in the runtime environment executing the query.

Reviewer: ```suggestion
indicates that the JDBC driver used by the ExaLoader successfully parsed the connection string but could not open the truststore because the specified path does not exist in the runtime environment executing the `IMPORT` query.

Note that this has nothing to do with the Virtual Schema adapter. There the JDBC configuration is correct.
```

Without SSL, no truststore is required, so the connection succeeds.

> When SSL is enabled, the JDBC driver attempts to load the truststore before establishing the TLS connection. Since the truststore cannot be resolved, the connection fails before any TLS handshake with Hive begins.

Reviewer: ```suggestion
When SSL is enabled, the JDBC driver in the loader attempts to load the truststore before establishing the TLS connection. Since the truststore cannot be resolved, the connection fails before any TLS handshake with Hive begins.
```

## Why metadata works but queries fail

Creating and using a Hive Virtual Schema involves two different phases:

1. Metadata access
2. Data access

Although both phases use the same connection definition, they are executed by different components inside Exasol.

### Metadata access

> During `CREATE VIRTUAL SCHEMA`, the Java adapter connects to Hive to retrieve metadata such as schemas, tables, and column definitions.

Reviewer: ```suggestion
During `CREATE VIRTUAL SCHEMA`, the virtual schema adapter (a UDF) connects to Hive to retrieve metadata such as schemas, tables, and column definitions.
```
